pkg/isb/serde.go

```diff
@@ -1,431 +1,126 @@
+	pb := &isb.Header{
+		MessageInfo: &isb.MessageInfo{
+			EventTime: timestamppb.New(h.MessageInfo.EventTime),
+			IsLate:    h.MessageInfo.IsLate,
+		},
+		Kind:    isb.MessageKind(h.Kind),
+		Id:      &isb.MessageID{VertexName: h.ID.VertexName, Offset: h.ID.Offset, Index: h.ID.Index},
+		Keys:    h.Keys,
+		Headers: h.Headers,
 	}
-	if err = binary.Write(buf, binary.LittleEndian, []byte(id.VertexName)); err != nil {
-		return nil, err
-	}
-	if err = binary.Write(buf, binary.LittleEndian, []byte(id.Offset)); err != nil {
-		return nil, err
-	}
-	return buf.Bytes(), nil
+	return proto.Marshal(pb)
 }
 
-// UnmarshalBinary decodes MessageID from the binary format
-func (id *MessageID) UnmarshalBinary(data []byte) (err error) {
-	var r = bytes.NewReader(data)
-	var preamble = new(messageIDPreamble)
-	if err = binary.Read(r, binary.LittleEndian, preamble); err != nil {
-		return err
-	}
-	var vertexName = make([]byte, preamble.VertexNameLen)
-	if err = binary.Read(r, binary.LittleEndian, vertexName); err != nil {
-		return err
-	}
-	var offset = make([]byte, preamble.OffsetLen)
-	if err = binary.Read(r, binary.LittleEndian, offset); err != nil {
+// UnmarshalBinary decodes Header from the proto bytes.
+func (h *Header) UnmarshalBinary(data []byte) error {
+	pb := &isb.Header{}
+	if err := proto.Unmarshal(data, pb); err != nil {
 		return err
 	}
-	id.VertexName = string(vertexName)
-	id.Offset = string(offset)
-	id.Index = preamble.Index
-	return nil
-}
 
-type headerPreamble struct {
-	// message length
-	MLen    int32
-	MsgKind MessageKind
-	// message ID len
-	IDLen int16
-	// message keys length
-	KeysLen int16
-	// header length
-	HeadersLen int16
-}
-
-// MarshalBinary encodes Header to a binary format
-func (h Header) MarshalBinary() (data []byte, err error) {
-	var buf = new(bytes.Buffer)
-	msgInfo, err := h.MessageInfo.MarshalBinary()
-	if err != nil {
-		return nil, err
-	}
-	id, err := h.ID.MarshalBinary()
-	if err != nil {
-		return nil, err
-	}
-	var preamble = headerPreamble{
-		MLen:       int32(len(msgInfo)),
-		MsgKind:    h.Kind,
-		IDLen:      int16(len(id)),
-		KeysLen:    int16(len(h.Keys)),
-		HeadersLen: int16(len(h.Headers)),
-	}
-	if err = binary.Write(buf, binary.LittleEndian, preamble); err != nil {
-		return nil, err
-	}
-	n, err := buf.Write(msgInfo)
-	if err != nil {
-		return nil, err
-	} else if n != int(preamble.MLen) {
-		return nil, fmt.Errorf("expected to write msgInfo size of %d but got %d", preamble.MLen, n)
-	}
-	n, err = buf.Write(id)
-	if err != nil {
-		return nil, err
-	} else if n != int(preamble.IDLen) {
-		return nil, fmt.Errorf("expected to write id size of %d but got %d", preamble.IDLen, n)
-	}
-	for i := 0; i < len(h.Keys); i++ {
-		if err = binary.Write(buf, binary.LittleEndian, int16(len(h.Keys[i]))); err != nil {
-			return nil, err
-		}
-		if err = binary.Write(buf, binary.LittleEndian, []byte(h.Keys[i])); err != nil {
-			return nil, err
-		}
+	h.MessageInfo = MessageInfo{
```

Comment:
```suggestion
	// no nil checks are performed because Marshalling and Unmarshalling
	// are completely controlled by numaflow. It is best not to do nil checks
	// and get panics so we can fix the root cause.
	
	h.MessageInfo = MessageInfo{
```